### PR TITLE
Migrate the settings page to Tailwind UI and Blade UI Kit

### LINKED ISSUE #576: Inform Twitter handles should not include @
On the profile settings page, the user now has the ability to add their Twitter. This should not include the `@` symbol and that is not currently obvious.

### PR DESCRIPTION
**Pull request by joedixon**
I looked through the settings pages of Tailwind UI and  like the layout I have opted for. This did mean merging the profile and the password settings onto one page (a separate form and panel for each).

Screenshot 2020-12-10 at 15 01 00

Screenshot 2020-12-10 at 15 03 05

Screenshot 2020-12-10 at 15 02 54

Screenshot 2020-12-10 at 15 02 00

Screenshot 2020-12-10 at 15 01 54

Screenshot 2020-12-10 at 15 01 08

Screenshot 2020-12-10 at 15 03 46

Also resolves #576

### CODE REVIEW

**Comment by driesvints** [comment]
This looks super good! But I'm thinking we can ditch the menu on the left entirely because it's already present in the navigation dropdown. I don't feel it serves much purpose there.

Sort of like this screen here: https://tailwindui.com/components/application-ui/page-examples/settings-screens#component-c29fe3f9f780c97a58f08453fc6bbec8

**Comment by joedixon** [comment]
> This looks super good! But I'm thinking we can ditch the menu on the left entirely because it's already present in the navigation dropdown. I don't feel it serves much purpose there.
> 
> Sort of like this screen here: https://tailwindui.com/components/application-ui/page-examples/settings-screens#component-c29fe3f9f780c97a58f08453fc6bbec8

My thinking was that as we add more user features, the navbar will get a bit bloated. I liked the idea of having a single navbar item that takes you do the user area and you can nvaigate your way from there.

I'm happyt to be talked around though 😄

**Comment by driesvints** [comment]
It just feels really out of place to me. Especially since the menu will disappear if you click any item except the settings. The navbar menu just has a handful of items and isn't bloated at all I think. Now that we have a single page for all our user settings I think it makes sense to just have the page centered without a menu next to it. If we have lots of extra settings in the future we can revise of course.

Screenshot 2020-12-11 at 15 41 28

**Comment by joedixon** [comment]
> It just feels really out of place to me. Especially since the menu will disappear if you click any item except the settings. The navbar menu just has a handful of items and isn't bloated at all I think. Now that we have a single page for all our user settings I think it makes sense to just have the page centered without a menu next to it. If we have lots of extra settings in the future we can revise of course.

My plan was to make the sidebar persistent across all user pages so in the long-run, it wouldn't disappear. I'm happy to park that idea for now and go with the single centered column. Not sure when I'll be able to get around to it, but will try and find some time next week.

**Comment by driesvints** [comment]
How would that look for the profile page though? What happens for the profile page when you're not logged in?

I think it's best to remove it for now and revisit it when we actually need it.

**Comment by joedixon** [comment]
Settings screenshot

I removed the sidebar and made the panels centered on the page. I also added the ability to have a prefix icon in the form input.

**Comment by driesvints** [comment]
@joedixon cool. Tests are failing it seems.

**Comment by joedixon** [comment]
> @joedixon cool. Tests are failing it seems.

That will teach me to change the button text. Should be good now.

File: app/View/Components/DangerButton.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    public function __construct(?string $type = 'button')
+    {
+        $this->type = $type;
+    }
+
+    public function render()
+    {
+        return view('components.buttons.danger-button');
+    }
+}
```
Comment (driesvints) [inline app/View/Components/DangerButton.php:20]: Does this need to be a class based component? Can just be an anonymous component imo.

File: app/View/Components/PrimaryButton.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    public function __construct(?string $type = 'button')
+    {
+        $this->type = $type;
+    }
+
+    public function render()
+    {
+        return view('components.buttons.primary-button');
+    }
+}
```
Comment (driesvints) [inline app/View/Components/PrimaryButton.php:20]: Does this need to be a class based component? Can just be an anonymous component imo.

File: app/View/Components/SidebarLink.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        $this->label = $label;
+        $this->icon = $icon;
+        $this->active = $active;
+    }
+
+    public function render()
+    {
+        return view('components.sidebar-link');
+    }
+}
```
Comment (driesvints) [inline app/View/Components/SidebarLink.php:29]: Does this need to be a class based component? Can just be an anonymous component imo.

File: config/blade-ui-kit.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        'mapbox' => Components\Maps\Mapbox::class,
+        'markdown' => Components\Markdown\Markdown::class,
+        'password' => Components\Forms\Inputs\Password::class,
+        'pikaday' => Components\Forms\Inputs\Pikaday::class,
+        'social-meta' => Components\Layouts\SocialMeta::class,
+        'textarea' => Components\Forms\Inputs\Textarea::class,
+        'toc' => Components\Markdown\ToC::class,
+        'trix' => Components\Editors\Trix::class,
+        'unsplash' => Components\Support\Unsplash::class,
+    ],
```
Comment (driesvints) [inline config/blade-ui-kit.php:45]: Let's remove all the components here which we don't use atm. Or comment them out.

File: resources/css/forms.css [hunk 5]
```diff
@@ -1,104 +1,104 @@
...
 
-.choices[data-type*='select-one']:after {
+.standard .choices[data-type*='select-one']:after {
     border-color: theme('colors.gray.500') transparent transparent transparent !important;
 }
 
-.choices[data-type*='select-one'] .choices__input {
+.standard .choices[data-type*='select-one'] .choices__input {
     @apply rounded-none border-b-2 border-gray-300 !important;
 }
```
Comment (driesvints) [inline resources/css/forms.css:104]: Could have just wrapped the place where the file is imported in a `.standard` class heh.

File: resources/helpers.php [hunk 6]
```diff
@@ -3,13 +3,27 @@
 if (! function_exists('active')) {
     /**
      * Sets the menu item class for an active route.
+     *
+     * @param mixed $routes
```
Comment (driesvints) [inline resources/helpers.php:7]: Remove this param. There's no type hint which already hints at it being mixed.

File: resources/helpers.php [hunk 7]
```diff
@@ -3,13 +3,27 @@
...
     {
         return call_user_func_array([app('router'), 'is'], (array) $routes) && $condition ? 'active' : '';
     }
 }
 
+if (! function_exists('is_active')) {
+    /**
+     * Determines if the given routes are active.
+     *
+     * @param mixed $routes
```
Comment (driesvints) [inline resources/helpers.php:19]: Remove this param. There's no type hint which already hints at it being mixed.

File: resources/helpers.php [hunk 8]
```diff
@@ -23,6 +37,8 @@ function md_to_html(string $markdown): string
 if (! function_exists('route_to_reply_able')) {
     /**
      * Returns the route for the replyAble.
+     *
+     * @param mixed $replyAble
```
Comment (driesvints) [inline resources/helpers.php:41]: Remove this param. There's no type hint which already hints at it being mixed.

File: resources/views/components/inputs/input.blade.php [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    <div class="relative">
+        <input
+            name="{{ $name }}"
+            type="{{ $type }}"
+            id="{{ $id }}"
+            @if($value)value="{{ $value }}"@endif
+            {{ $attributes }}
+            class="{{ $class }} {{ $errors->has($name) ? $errorClasses() : '' }}"
+        />
+        @if($errors->has($name))
```
Comment (driesvints) [inline resources/views/components/inputs/input.blade.php:11]:
```suggestion

        @if ($errors->has($name))
```

File: resources/views/components/inputs/input.blade.php [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+            <div class="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none">
+                <x-heroicon-o-exclamation-circle class="h-5 w-5 text-red-500" />
+            </div>
+        @endif
+    </div>
+    @if($errors->has($name))
```
Comment (driesvints) [inline resources/views/components/inputs/input.blade.php:17]:
```suggestion

    @if ($errors->has($name))
```

File: resources/views/users/settings/profile.blade.php [hunk 11]
```diff
@@ -1,80 +1,79 @@
...
-            @formGroup('twitter')
-                <label for="twitter">Twitter handle</label>
-                <input type="text" name="twitter" id="twitter" value="{{ Auth::user()->twitter() }}" />
-                @error('twitter_handle')
-            @endFormGroup
+                <div class="grid grid-cols-12 gap-6">
+                    <div class="col-span-12">
+                        <x-label for="email" />
+                        <x-input type="email" name="email" value="{{ Auth::user()->emailAddress() }}" required />
+                        @unless(Auth::user()->hasVerifiedEmail())
```
Comment (driesvints) [inline resources/views/users/settings/profile.blade.php:45]:
```suggestion

                        @unless(Auth::user()->hasVerifiedEmail())
```

File: resources/views/users/settings/remove.blade.php [hunk 12]
```diff
@@ -0,0 +1,27 @@
+@unless (Auth::user()->isAdmin())
+    <section aria-labelledby="remove_account_heading" class="mt-6">
+        <div class="shadow sm:rounded-md sm:overflow-hidden">
+            <div class="bg-white py-6 px-4 space-y-6 sm:p-6">
+                <div>
+                    <h2 id="remove_account_heading" class="text-lg leading-6 font-medium text-red-500 uppercase">Danger Zone</h2>
+                    <p class="mt-1 text-sm leading-5 text-gray-500">Please be aware that deleting your account will also remove all of your data, including your threads and replies. This cannot be undone.</p>
```
Comment (driesvints) [inline resources/views/users/settings/remove.blade.php:7]:
```suggestion
                    <p class="mt-1 text-sm leading-5 text-gray-500">
                        Please be aware that deleting your account will also remove all of your data, including your threads and replies. This cannot be undone.
                    </p>
```

File: resources/views/vendor/blade-ui-kit/components/forms/inputs/input.blade.php [hunk 13]
```diff
@@ -0,0 +1,28 @@
+<div>
+    <div class="relative">
+        @if ($attributes->get('prefix-icon'))
+            <div class="absolute inset-y-0 left-0 pl-3 flex items-center pointer-events-none">
+                @svg($attributes->get('prefix-icon'), ['class' => 'h-5 w-5 text-gray-400'])
+            </div>
+        @endif
+        <input
```
Comment (driesvints) [inline resources/views/vendor/blade-ui-kit/components/forms/inputs/input.blade.php:8]:
```suggestion

        <input
```

File: resources/views/vendor/blade-ui-kit/components/forms/inputs/input.blade.php [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+            name="{{ $name }}"
+            type="{{ $type }}"
+            id="{{ $id }}"
+            @if($value)value="{{ $value }}"@endif
+            {{ $attributes->merge([
+                'class' => 'form-input block w-full sm:text-sm sm:leading-5 mt-1' . ($attributes->get('prefix-icon') ? ' pl-10' : '') . ($errors->has($name) ? ' border-red-300 text-red-900 placeholder-red-300 focus:outline-none focus:ring-red-500 focus:border-red-500' : '')
+            ]) }}"
+        />
+        @if ($errors->has($name))
```
Comment (driesvints) [inline resources/views/vendor/blade-ui-kit/components/forms/inputs/input.blade.php:17]:
```suggestion

        @if ($errors->has($name))
```

File: resources/views/vendor/blade-ui-kit/components/forms/inputs/input.blade.php [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+            <div class="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none">
+                <x-heroicon-o-exclamation-circle class="h-5 w-5 text-red-500" />
+            </div>
+        @endif
+    </div>
+    @if ($errors->has($name))
```
Comment (driesvints) [inline resources/views/vendor/blade-ui-kit/components/forms/inputs/input.blade.php:23]:
```suggestion

    @if ($errors->has($name))
```

File: resources/views/vendor/blade-ui-kit/components/forms/inputs/textarea.blade.php [hunk 16]
```diff
@@ -0,0 +1,20 @@
+<div class="relative">
+    <textarea
+        name="{{ $name }}"
+        id="{{ $id }}"
+        rows="{{ $rows }}"
+        {{ $attributes->merge([
+            'class' => 'form-textarea block w-full transition duration-150 ease-in-out sm:text-sm sm:leading-5' . ($errors->has($name) ? ' border-red-300 text-red-900 placeholder-red-300 focus:outline-none focus:ring-red-500 focus:border-red-500' : '')
+        ]) }}"
+    >{{ old($name, $slot) }}</textarea>
+    @if($errors->has($name))
```
Comment (driesvints) [inline resources/views/vendor/blade-ui-kit/components/forms/inputs/textarea.blade.php:10]:
```suggestion

    @if($errors->has($name))
```

File: resources/views/vendor/blade-ui-kit/components/forms/inputs/textarea.blade.php [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+        <div class="absolute inset-y-0 right-0 pr-3 flex items-center pointer-events-none">
+            <x-heroicon-o-exclamation-circle class="h-5 w-5 text-red-500" />
+        </div>
+    @endif
+    @if($errors->has($name))
```
Comment (driesvints) [inline resources/views/vendor/blade-ui-kit/components/forms/inputs/textarea.blade.php:15]:
```suggestion

    @if($errors->has($name))
```

**Review by driesvints — CHANGES_REQUESTED** [review]
Nice job. I left some remarks.

A few notes in general:
- Always try to leave some breathing room before or after an if directive in Blade views when there's elements on the same level before or after it.
- If directives always have a space between `if` and the opening `(`
- I see a lot of `{{` syntax being used in Blade component value attributes. Did you test all of these? Because afaik this isn't supported on Blade components. You should try to use the `:value="Auth::user()->email"` syntax instead.

**Comment by driesvints** [comment]
@joedixon tests seem to be failing.

**Comment by joedixon** [comment]
> @joedixon tests seem to be failing.

Fixed!

**Review by driesvints — APPROVED** [review]

**Pull request merged**